src/main/java/com/softserve/bookstore/repositories/UserRepository.java

```diff
@@ -2,123 +2,148 @@
+        Connection connection = connectionManager.getConnection();
+        PreparedStatement userStatement = connection.prepareStatement(SELECT_LAST_USERS);
+        userStatement.setInt(1, numberOfRecords);
+        ResultSet resultSet = userStatement.executeQuery();
+        List<User> users = getUsersFromResultSet(resultSet);
+        Collections.reverse(users);
+        return users;
+    }
+
+    private  void addUserToBatch(PreparedStatement userStatement, User user) {
+        try {
+            userStatement.setString(1, user.getEmail());
+            userStatement.setString(2, user.getPassword());
+            userStatement.addBatch();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
 
-    public Optional<User> getUserByEmail(List<User> users, String email) {
+    public Optional<User> getUserByEmail(List<User> users, String email){
         return users.stream()
                 .filter(user -> email.equals(user.getEmail().trim()))
                 .findFirst();
     }
 
-    public void add(List<User> users) throws SQLException {
+    public void addUsers(List<User> users) throws SQLException{
         Connection connection = connectionManager.getConnection();
         PreparedStatement userStatement = connection.prepareStatement(INSERT_USERS);
         users.forEach(user -> {
-            UserUtility.addUserToBatch(userStatement, user);
+            addUserToBatch(userStatement, user);
         });
         userStatement.executeBatch();
 
         List<User> lastUsersAdded = findLastUsersAdded(users.size());
         lastUsersAdded.forEach(user -> {
             try {
-                roleRepository.addRole(Role.USER.toString(), user);
-                getUserByEmail(users, user.getEmail())
-                        .orElseThrow(() -> new UserNotFoundException("User does not exist."));
-
-                User currentUser = getUserByEmail(users, user.getEmail()).get();
-
-                if (currentUser.getRoles().size() > 1) {
-                    roleRepository.addRole(Role.ADMIN.toString(), user);
-                }
-                if (currentUser.getOrders().size() > 1) {
-                    addOrdersToUser(currentUser, users.size());
+                    addRole(Role.USER.toString(), user);
+                    getUserByEmail(users, user.getEmail())
+                            .orElseThrow(() -> new UserNotFoundException("User with email"));
+
+                    User currentUser = getUserByEmail(users, user.getEmail()).get();
+
+                    if (currentUser.getRoles().size() > 1) {
+                        addRole(Role.ADMIN.toString(), user);
+                    }
+                    if (currentUser.getOrders().size() > 1) {
+                        addOrdersToUser(currentUser, users.size());
+                    }
+                } catch (UserNotFoundException | SQLException e) {
+                    throw new RuntimeException(e);
```

Comment:
For exception handling just have single point where exception will be handled,
in most cases it is done at the level of controller layer (web service layer), where exception will be handled
exception message will be included in response with proper response, like 404, 503, etc.
and also logging should be done at that point of exception handling

Reply:
here is also a useful article about exception handling best practices in java
https://stackify.com/best-practices-exceptions-java/
consider especially the point about preferring specific exceptions and descriptive message
so in this case just let SqlException be thrown,
if it is thrown than it will be handled at the controller layer (or somewhere else)